File: Fore-C_v2/codes/Models_save_best.R
# Re-run best models, save objects
library(quantregForest)

# load best models key
best_models_final <- read.csv('../model_selection_summary_results/parsimonious_best_models_by_disease_and_region.csv')

# directory paths
smote_dir <- '../compiled_data/survey_data/smote_datasets/'
model_objects_dir <- '../model_objects/'

# run models and save model objects
for(i in 1:nrow(best_models_final)){
  # format filenames for best models
  thsh <- ifelse(nchar(best_models_final$threshold[i]) == 2, best_models_final$threshold[i], paste0('0', best_models_final$threshold[i]))
  dz_name <- ifelse(best_models_final$name[i] == 'ws_pac', 'ws_pac_acr', best_models_final$name[i])
  train_name <- paste0(smote_dir, dz_name, '_smote_train_', thsh, '.csv')
  # test_name <- gsub('train', 'test', train_name)
  
  # load data
  df_train <- read.csv(train_name)
  # df_test <- read.csv(test_name)

  # format data for quantile regression
  dz_vars <- unlist(strsplit(best_models_final$Model_variables[i], split=', '))
  yVar <- ifelse(dz_name == 'ga_gbr' | dz_name == 'ws_gbr', 'Y', 'p')
  
  x_train <- df_train[, dz_vars] 
  y_train <- df_train[, yVar]

  # x_test <- df_test[, dz_vars] 
  # y_test <- df_test[, yVar]
  
  # run best model
  final_mod <- quantregForest(x_train, y_train, importance = TRUE)
  
  # save model
  modFileName <- paste0(model_objects_dir, best_models_final$name[i], ".rds")
  saveRDS(final_mod, file = modFileName)
}

File: Fore-C_v2/codes/Models_save_best.py
# -*- coding: utf-8 -*-
"""
Created on Fri May 20 12:53:12 2022

@author: jamie
"""

# in terminal
# pip install setuptools numpy scipy scikit-learn cython
# pip install scikit-garden

import pandas as pd
from sklearn.ensemble import RandomForestRegressor
import joblib

# semi-hard code initially
x = pd.read_csv('../model_selection_summary_results/parsimonious_best_models_by_disease_and_region.csv')

# update disease-region name
x = x.replace('ws_pac', 'ws_pac_acr')

# change threshold from '5' to '05'
x.threshold = x.threshold.replace(5, '05')


# set where model objects will be stored
model_objects_dir = "../model_objects/"

# create and save models
for i in range(4):
    # load dataset
    train_df_filename = str('../compiled_data/survey_data/smote_datasets/') + x.name[i] + str('_smote_train_') + str(x.threshold[i]) + str('.csv')
    train_df = pd.read_csv(train_df_filename)

    # determine response variables based on whether or not column names include "p"
    if 'p' in train_df.columns:
        response = 'p'
    else:
        response = 'Y'
        
    # list covariates
    df_covars = x.iloc[i].Model_variables.split(", ")
    
    # format data for quantile forecst regression
    X_train = train_df[df_covars] 
    Y_train = train_df[response]
    
    # create model
    # n_estimators is number of trees, goes slower as number goes up
    qrf = RandomForestRegressor(n_estimators = 200, random_state = 0, min_samples_split = 10)
    qrf.fit(X_train, Y_train)

    # save models
    fileName = model_objects_dir + x.name[i] + str('.joblib')
    joblib.dump(qrf, fileName)
    
    # plots: not essential, so comment in/out as desired
    # import matplotlib.pyplot as plt
    # from sklearn.inspection import PartialDependenceDisplay
    fig, ax = plt.subplots(figsize=(14, 10))
    ax.set_title(x.name[i])
    pdpfig = PartialDependenceDisplay.from_estimator(qrf, X_train, df_covars, ax = ax)

#     fig2, ax2 = plt.subplots(figsize=(14, 10))
#     ax2.set_title(x.name[i])
#     pdp_plot2 = PartialDependenceDisplay.from_estimator(qrf, X_train, df_covars, ax = ax2, kind = 'both')

    # variable importance plot (https://scikit-learn.org/stable/auto_examples/ensemble/plot_forest_importances.html)
    importances = qrf.feature_importances_
    std = np.std([tree.feature_importances_ for tree in qrf.estimators_], axis=0)

    importances = qrf.feature_importances_
    feature_names = qrf.feature_names_in_
    forest_importances = pd.Series(importances, index=feature_names)

    fig, ax = plt.subplots()
    forest_importances.plot.bar(yerr = std, ax = ax)
    ax.set_title("Feature importances using MDI")
    ax.set_ylabel("Mean decrease in impurity")
    fig.tight_layout()
    fig_filepath = str('../../Figures/Quantile_forests/variable_importance/') + x.name[i] + str('.pdf')
    fig.savefig(fig_filepath) 



